Reduce number of `JOIN`s in builds query

aoldershaw opened the pull request:
## Changes proposed by this PR

* [x] Don't `JOIN` on `resource_types` or `prototypes` when `SELECT`ing a build
  * The reasoning is explained more thoroughly in add3a0b, but tl;dr, we were exceeding the default `join_collapse_limit` (of 8 relations) - if you exceed this limit, then Postgres refuses to optimize the order in which it performs the `JOIN`s, so you end up with a (potentially very) suboptimal execution plan
    * When we deployed to prod, the normal "get a check build" query that normally takes <100ms was taking >5 minutes 😬 
  * We previously performed the `JOIN`s to label builds as for a particular resource type or prototype, by name. Now, builds specifically for a resource type or prototype will appear as one-off builds (in tracing, logging, and the syslog drainer)
  * With #7176, we don't generate check builds for `resource_types` (or `prototypes`), so this will only apply to `fly check-resource-type`/`-prototype`
* [x] Add a wrapper `db.Conn` to fail our test suite if we introduce a query with "too many" `JOIN`s

aoldershaw committed:
B: avoid hitting join_collapse_limit

adding prototypes to the "scan build" query pushed the number of joins
over the join_collapse_limit (defaults to 8) for certain queries,
causing postgres to generate suboptimal execution plans (taking
~5minutes for a simple query, since it ended up doing seq scans for all
tables).

this commit removes the joins for resource_types and prototypes. these
joins were just to get the names of the resource_type/prototype for use
in the syslog tag, tracing, and logs. we no longer add this information
to these sinks. however, with @clarafu's refactor in #7176, we won't
generate builds for resource_types or prototypes automatically (i.e. in
lidar), as they will be checked as part of the checks for resources (or
steps in a job build), so hopefully this isn't too big a deal.

in the long run, a better solution may be to store general metadata on
the builds table so we don't have to do so many joins in the first place
- but for now, let's just get rid of these less interesting ones to
exceeding the limit

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Alex Suraci <suraci.alex@gmail.com>

aoldershaw committed:
add sanity check for joining to too many tables

this commit adds a wrapper over db.Conn that ensures we don't have more
than 8 JOINs in a single query, so we can catch performance issues
without needing to deploy

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Alex Suraci <suraci.alex@gmail.com>

chenbh commented on atc/postgresrunner/join_limit_conn.go:
```diff
@@ -0,0 +1,96 @@
+
+func validateJoinLimit(query string) error {
+	// Each subquery, split by a UNION or a UNION ALL, can have up to 8 JOINs
+	subQueries := strings.Split(query, "UNION")
```
nit: should we uppercase the query here incase we use lowercase `union`?

chenbh commented on atc/postgresrunner/join_limit_conn.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+func validateJoinLimit(query string) error {
+	// Each subquery, split by a UNION or a UNION ALL, can have up to 8 JOINs
```
Should we split out CTEs as well? I'm assuming since they're executed independently of the main query, they shouldn't contribute to the limit
```
WITH foo AS (
SELECT ... JOIN ... JOIN ... JOIN ... JOIN ...JOIN
),
SELECT ... JOIN ... JOIN ... JOIN ... JOIN
```

aoldershaw replied:
Added cd1c046f70b5f741db93e75bbd60b8acc4f48bef that addresses CTEs and any subqueries (I don't *think* subqueries count toward the `JOIN` limit...). Happy to revert it as it's probably too complex for the benefit it adds - I was mostly just having fun with it.

Honestly, keeping the naïve `UNION` only splitting maybe isn't such a big deal, since we're at least guaranteed there's no false negatives.

aoldershaw committed:
make query parsing more robust

this adds an query extraction method to reduce the likelihood of false
positives by treating CTEs, UNIONs, and subqueries as separate when
cointing the number of JOINs

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

chenbh commented:
Do you know why it defaults to 8? While I think it's fine to get rid of the 2 joins for this particular query, it might be interesting to see how much performance hit we take if we bump the `join_collapse_limit` (and `from_collapse_limit` as well) up instead.

aoldershaw commented:
@chenbh we considered just bumping the `join_collapse_limit`, but decided against it for now just because there was this relatively low-hanging fruit.

The higher the `join_collapse_limit`, the longer planning will take. At least theoretically, if the limit is `n`, Postgres considers up to `n!` orderings, which can grow pretty quickly. That said, I'm guessing it doesn't actually need to consider all `n!` possibilities.

That said, if we continue to run into this problem (which hopefully we'll be able to detect now), we may consider bumping the limit.

chenbh reviewed (approved):
logic lgtm, I'll defer to @clarafu in case she has any db optimizations/tricks to help with these types of scenarios

aoldershaw commented:
@chenbh going to merge this for now since this is a blocker for deploying to CI. If you get the chance, can you rebase causality on to this? Curious to see how performant the `version_md5`-based algorithm is in practice.

chenbh commented:
done!
